pkg/validation/app.go

```diff
@@ -97,6 +103,58 @@ func (v *Validator) validateConfig(ctx context.Context, cr v1alpha1.App) error {
 	return nil
 }
 
+func (v *Validator) validateMetadataConstraint(ctx context.Context, cr v1alpha1.App) error {
+	name := key.AppCatalogEntryName(key.CatalogName(cr), key.AppName(cr), key.Version(cr))
+
+	entry, err := v.g8sClient.ApplicationV1alpha1().AppCatalogEntries(metav1.NamespaceDefault).Get(ctx, name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		v.logger.Debugf(ctx, "appCatalogEntry %#q not found, skipping validation", name)
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if entry.Spec.Restrictions == nil {
+		// no-op
+		return nil
+	}
+
+	if entry.Spec.Restrictions.FixedNamespace != "" {
+		if entry.Spec.Restrictions.FixedNamespace != cr.Spec.Namespace {
+			return microerror.Maskf(validationError, "app %#q could be installed in namespace %#q only, not %#q",
+				cr.Spec.Name, entry.Spec.Restrictions.FixedNamespace, cr.Spec.Namespace)
+		}
+	}
+
+	var apps *v1alpha1.AppList
+	if entry.Spec.Restrictions.ClusterSingleton || entry.Spec.Restrictions.NamespaceSingleton {
+		lo := metav1.ListOptions{
+			FieldSelector: fmt.Sprintf("metadata.name!=%s", cr.Name),
+		}
+		apps, err = v.g8sClient.ApplicationV1alpha1().Apps(cr.Namespace).List(ctx, lo)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	for _, app := range apps.Items {
+		if app.Spec.Name == cr.Spec.Name {
+			if entry.Spec.Restrictions.ClusterSingleton {
+				return microerror.Maskf(validationError, "app %#q could be installed only once in cluster %#q",
+					cr.Spec.Name, cr.Namespace)
+			}
+			if entry.Spec.Restrictions.NamespaceSingleton {
+				if app.Spec.Namespace == cr.Spec.Namespace {
```

Comment:
When `namespace` is the same between two objects, it's violating the `namespaceSingleton` restriction.